File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.py
class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        N, result = len(nums), 0
        max_win, min_win = deque([]), deque([])
        l = 0
        for r in range(N):
            while max_win and nums[max_win[-1]] < nums[r]:
                max_win.pop()

            while min_win and nums[min_win[-1]] > nums[r]:
                min_win.pop()
            
            max_win.append(r)
            min_win.append(r)
            while nums[max_win[0]] - nums[min_win[0]] > limit:
                if max_win[0] < min_win[0]:
                    l = max(l, max_win[0] + 1)
                    max_win.popleft()
                else:
                    l = max(l, min_win[0] + 1)
                    min_win.popleft()
            result = max(result, r - l + 1)
        
        return result

File: 1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit/1438-longest-continuous-subarray-with-absolute-diff-less-than-or-equal-to-limit.cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums, int limit) {
        int N = nums.size(), result = 0;
        deque<int> max_win, min_win;
        for(int r=0, l=0; r<N; r++){
            while(!max_win.empty() && nums[max_win.back()] < nums[r]){
                max_win.pop_back();
            }
            while(!min_win.empty() && nums[min_win.back()] > nums[r]){
                min_win.pop_back();
            }   
            max_win.push_back(r);
            min_win.push_back(r);
            while(nums[max_win.front()] - nums[min_win.front()] > limit){
                if(max_win.front() < min_win.front()) {
                    l = max(l, max_win.front() + 1);
                    max_win.pop_front();
                }
                else {
                    l = max(l, min_win.front() + 1);
                    min_win.pop_front();
                }
            }
            result = max(result, r - l + 1);
        }
        return result;
    }
};